### PR TITLE
refactor: Refactor station finder to support multiple strategy

### PR DESCRIPTION
**Pull request by CodeBear801**
issue: https://github.com/Telenav/osrm-backend/issues/270

- Refactor code inside `stationfinder`
- Add interface of finder which hides implementation of tnsearch or local search
```go
finder := CreateStationsFinder(findertype, ...) 

// ...

origStations := finder.NewOrigStationFinder(req)
destStations := finder.NewDestStationFinder(req)
overlap := stationfinder.FindOverlapBetweenStations(origStations, destStations)
```
- Create inteface.go to records all external interfaces.
- Create package `stationfindertype` to records all output related data structures
   + which is also needed by implementations, I use another package to hold them to avoid cycle dependencies
- `tnsearchfinder` implements finder based on Telenav search
   + later will implment `localsearchfinder` based on google:s2
- Change `cost` to `weight`
- Change related name in `stationgraph`

## Tasklist

 - [ ] refactor code
 - [ ] review

### CODE REVIEW

File: integration/service/oasis/stationfinder/tnsearchfinder/low_energy_location_station_finder_test.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+	"github.com/Telenav/osrm-backend/integration/service/oasis/stationfinder/stationfindertype"
+)
+
+func TestLowEnergyLocationStationFinderIterator1(t *testing.T) {
+	sf := createMockLowEnergyLocationStationFinder1()
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func(wg *sync.WaitGroup) {
```
Comment (wangyoucao577) [inline integration/service/oasis/stationfinder/tnsearchfinder/low_energy_location_station_finder_test.go:16]: Nit: It's a simple sync funciton, the `go func() ...` and `waitgroup` seems redundant.

File: integration/service/oasis/stationfinder/tnsearchfinder/orig_station_finder_test.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+	"testing"
+
+	"github.com/Telenav/osrm-backend/integration/service/oasis/stationfinder/stationfindertype"
+)
+
+func TestOrigStationFinderIterator(t *testing.T) {
+	sf := CreateMockOrigStationFinder1()
+	c := sf.IterateNearbyStations()
+	var r []stationfindertype.ChargeStationInfo
+	go func() {
```
Comment (wangyoucao577) [inline integration/service/oasis/stationfinder/tnsearchfinder/orig_station_finder_test.go:14]: It's risky that run async inside but no wait.

File: integration/service/oasis/stationfinder/tnsearchfinder/tnsearch_station_finder.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
+package tnsearchfinder
```
Comment (wangyoucao577) [inline integration/service/oasis/stationfinder/tnsearchfinder/tnsearch_station_finder.go:1]: Nit: Can we remove the `tn`?

File: integration/service/oasis/stationfinder/tnsearchfinder/tnsearch_station_finder.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+	"github.com/Telenav/osrm-backend/integration/service/oasis/searchconnector"
+	"github.com/Telenav/osrm-backend/integration/service/oasis/stationfinder/stationfindertype"
+)
+
+type tnSearchStationFinder struct {
+	sc *searchconnector.TNSearchConnector
+}
+
+// NewTnSearchStationFinder creates finder based on telenav search
+func NewTnSearchStationFinder(sc *searchconnector.TNSearchConnector) *tnSearchStationFinder {
```
Comment (wangyoucao577) [inline integration/service/oasis/stationfinder/tnsearchfinder/tnsearch_station_finder.go:15]: Simply `New()`

File: integration/service/oasis/stationfinder/tnsearchfinder/tnsearch_station_finder.go (hunk 3 above)
Comment (CodeBear801) [reply integration/service/oasis/stationfinder/tnsearchfinder/tnsearch_station_finder.go:1]: Then would be two finders
- searchfinder
- localfinder

Let me think about it.  Or change to `cloudfinder`?

File: integration/service/oasis/stationfinder/tnsearchfinder/tnsearch_station_finder.go (hunk 4 above)
Comment (CodeBear801) [reply integration/service/oasis/stationfinder/tnsearchfinder/tnsearch_station_finder.go:15]: Copy that, package name already indicate for what to new

**Comment by CodeBear801** [comment]
Will make related changes in next pull.

File: integration/service/oasis/stationfinder/tnsearchfinder/low_energy_location_station_finder_test.go (hunk 1 above)
Comment (CodeBear801) [reply integration/service/oasis/stationfinder/tnsearchfinder/low_energy_location_station_finder_test.go:16]: do you mean using a `channel` to wait for return value instead?

Comment (wangyoucao577) [reply integration/service/oasis/stationfinder/tnsearchfinder/low_energy_location_station_finder_test.go:16]: No. The test function only includes a `go func()`. I mean, is it really need the async?

File: integration/service/oasis/stationfinder/tnsearchfinder/tnsearch_station_finder.go (hunk 3 above)
Comment (wangyoucao577) [reply integration/service/oasis/stationfinder/tnsearchfinder/tnsearch_station_finder.go:1]: How about `searchfinder` and `s2finder`?

File: integration/service/oasis/stationfinder/tnsearchfinder/low_energy_location_station_finder_test.go (hunk 1 above)
Comment (CodeBear801) [reply integration/service/oasis/stationfinder/tnsearchfinder/low_energy_location_station_finder_test.go:16]: Copy that, test function could be run in different go-routines, if just use one `go func()` then might be no need addition `no`

**Pull request merged**